### PR TITLE
Improve usability of the StrimziKafkaContainer

### PR DESCRIPTION
**Pull request by ozangunalp**
This PR aims to improve the usability of the Strimzi Kafka test container. This will lead us to use this Test container in Kafka related integration tests in Quarkus and Smallrye Reactive Messaging.

* Default values for kafka version and image version
* Variable for base image name: gives the ability to use strimzi image
* Fix for `waitForRunning` when not in kraft mode
* Option for setting fixed exposed port
* Option for setting server.properties
* Option for custom bootstrap servers: Gives way for SSL or SASL connections and also proxying the Kafka container.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -32,17 +32,17 @@ Or you can use `StrimziKafkaCluster` for a Kafka cluster with multiple Kafka bro
...
+2. Run Strimzi test container
     1. default configuration
     2. (optional) run Strimzi test container with additional configuration
 
-First, you need to specify two essential environment variables:
+You can specify environment variables to configure the container image to run:
 
-- `STRIMZI_TEST_CONTAINER_KAFKA_VERSION` - specification of the Kafka version to run with the given container. For example, `3.0.0`.
-- `STRIMZI_TEST_CONTAINER_IMAGE_VERSION` - specification of the Strimzi test container image version that was built in the `test-container-images` repositories. For example, `0.1.0`.
+- `STRIMZI_TEST_CONTAINER_KAFKA_VERSION` - the Kafka version to run with the given container. Default : `3.0.0`.
```
Comment (see-quick) [inline README.md:43]: We are not gonna have these ENVs about (test-container, Kafka version). In this PR https://github.com/strimzi/test-container/pull/7 I have replaced it with a mechanism that will fetch these properties and use it.

File: src/main/java/io/strimzi/utils/Environment.java [hunk 2]
```diff
@@ -12,15 +12,63 @@
     // empty constructor to prohibit instantiate object
     private Environment() { }
 
+    /**
+     * Strimzi test container base image environment variable
+     */
+    public static final String STRIMZI_TEST_CONTAINER_BASE_IMAGE_ENV = "STRIMZI_TEST_CONTAINER_BASE_IMAGE";
```
Comment (see-quick) [inline src/main/java/io/strimzi/utils/Environment.java:18]: Environment class will not exist anymore...(more on mentioned PR)

**Review by see-quick — COMMENTED** [review]
First of all thanks for the PR, it's great to have also specified deployment via server.properties file. 👍. I have a few comments regarding to my upcoming change... :)

**Comment by ozangunalp** [comment]
@see-quick Thanks for the feedback! Removing `Environment` class will be much appreciated. 
However, although I share the remark on many factory methods I don't see the need for a builder indirection. `GenericContainer` fluent methods already provide a builder pattern until the container is started.

I can rebase this PR on #7 as soon as you deem it as stable.

**Comment by see-quick** [comment]
@ozangunalp you can now rebase :)

**Comment by ozangunalp** [comment]
@see-quick Thanks! I had to take some things from scratch:

- Refactor usage of `KafkaVersionService` to fetch the versions from the configured URL only if kafka version and strimzi version are not provided.
- Remove compile dependencies to log implementations and also the log4j2 configuration included as resource.

File: README.md [hunk 3]
```diff
@@ -32,17 +32,17 @@ Or you can use `StrimziKafkaCluster` for a Kafka cluster with multiple Kafka bro
...
-2. Add the Strimzi test container as a Maven dependency
-3. Run Strimzi test container
+Using Strimzi test container takes two simple steps:
+1. Add the Strimzi test container as a Maven dependency
+2. Run Strimzi test container
     1. default configuration
     2. (optional) run Strimzi test container with additional configuration
 
-First, you need to specify two essential environment variables:
+You can specify environment variables to configure the container image to run:
```
Comment (see-quick) [inline README.md:41]: Did you rebase also `README.md`? It seems that you have a version with the `Env` stuff, which by this point is obsolete and we are no longer to use it :)

File: src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java [hunk 4]
```diff
@@ -98,19 +82,26 @@ protected void doStart() {
...
 
-        kafkaDynamicKafkaPort = getMappedPort(Constants.KAFKA_PORT);
+        this.kafkaExposedPort = getMappedPort(Constants.KAFKA_PORT);
 
-        LOGGER.info("Mapped port: {}", kafkaDynamicKafkaPort);
+        LOGGER.info("Mapped port: {}", kafkaExposedPort);
 
-        StringBuilder advertisedListeners = new StringBuilder(getBootstrapServers());
+        String bootstrapServers = getBootstrapServers();
+        String bsListenerName = extractListenerName(bootstrapServers);
```
Comment (see-quick) [inline src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java:102]: These two variables could be `final` or?

File: src/main/java/io/strimzi/test/container/utils/KafkaVersionService.java [hunk 5]
```diff
@@ -45,6 +50,37 @@ public KafkaVersionService() {
         this.resolveAndParse();
     }
 
+    /**
+     * Get singleton instance lazily
+     *
+     * @return singleton instance
+     */
+    public static KafkaVersionService getInstance() {
```
Comment (see-quick) [inline src/main/java/io/strimzi/test/container/utils/KafkaVersionService.java:58]: This is a good solution 👍 and replacement of `buildDefaults` method.

**Review by see-quick — COMMENTED** [review]
Mostly nits...Thanks for such contribution )

**Review by see-quick — APPROVED** [review]

File: README.md (hunk 3 above)
Comment (ozangunalp) [reply README.md:41]: It think they were there as I didn't have any conflicts. I am removing them

File: README.md [hunk 6]
```diff
@@ -99,7 +99,41 @@ StrimziKafkaContainer strimziKafkaContainer = new StrimziKafkaContainer()
...
+
+You can configure Kafka by providing a `server.properties` file:
+
+```java
+StrimziKafkaContainer strimziKafkaContainer = strimziKafkaContainer
+        .withServerProperties(MountableFile.forClasspathResource("server.properties"));
+strimziKafkaContainer.start();
+```
+
+Note that configuration properties `listeners`, `advertised.listeners`, `listener.security.protocol.map`, `inter.broker.listener.name`, `controller.listener.names`, `zookeeper.connect` and any additional configuration will be overridden during container startup.
```
[the remark this answers is not in the document]
Comment (ozangunalp) [reply README.md:116]: Indeed I need to rephrase this line

**Comment by ozangunalp** [comment]
@tombentley Thanks for the review, I fixed them on the last commit I pushed. Let me know what do you think

**Review by tombentley — APPROVED** [review]
Thanks!

**Pull request merged**